watcher/tool/watcher/alerts.go

```diff
@@ -0,0 +1,246 @@
+	"github.com/gravitational/monitoring-app/watcher/lib/kapacitor"
+	"github.com/gravitational/monitoring-app/watcher/lib/kubernetes"
+
+	"github.com/ghodss/yaml"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeapi "k8s.io/client-go/kubernetes"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+func runAlertsWatcher(kubernetesClient *kubernetes.Client) error {
+	kapacitorClient, err := kapacitor.NewClient()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	alertLabel, err := kubernetes.MatchLabel(constants.MonitoringLabel, constants.MonitoringUpdateAlert)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	targetLabel, err := kubernetes.MatchLabel(constants.MonitoringLabel, constants.MonitoringUpdateAlertTarget)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	smtpLabel, err := kubernetes.MatchLabel(constants.MonitoringLabel, constants.MonitoringUpdateSMTP)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	alertCh := make(chan map[string]string)
+	alertTargetCh := make(chan map[string]string)
+	configmaps := []kubernetes.ConfigMap{
+		{alertLabel, alertCh},
+		{targetLabel, alertTargetCh},
+	}
+	smtpCh := make(chan map[string][]byte)
+
+	go kubernetesClient.WatchConfigMaps(context.TODO(), configmaps...)
+	go kubernetesClient.WatchSecrets(context.TODO(), kubernetes.Secret{smtpLabel, smtpCh})
+	receiverLoop(context.TODO(), kubernetesClient.Clientset, kapacitorClient,
+		alertCh, alertTargetCh, smtpCh)
+
+	return nil
+}
+
+func receiverLoop(ctx context.Context, kubeClient *kubeapi.Clientset, kClient *kapacitor.Client,
+	alertCh, alertTargetCh <-chan map[string]string, smtpCh <-chan map[string][]byte) {
+	for {
+		select {
+		case update := <-alertCh:
+			spec := []byte(update[constants.ResourceSpecKey])
+			if err := createAlert(kClient, spec); err != nil {
+				log.Warnf("failed to create alert: %v", trace.DebugReport(err))
+			}
+		case update := <-smtpCh:
+			spec := update[constants.ResourceSpecKey]
+			client := kubeClient.Secrets(api.NamespaceSystem)
+			if err := updateSMTPConfig(client, kClient, spec); err != nil {
+				log.Warnf("failed to update SMTP configuration: %v", trace.DebugReport(err))
+			}
+		case update := <-alertTargetCh:
+			spec := []byte(update[constants.ResourceSpecKey])
+			client := kubeClient.ConfigMaps(api.NamespaceSystem)
+			if err := updateAlertTarget(client, kClient, spec); err != nil {
+				log.Warnf("failed to update alert target: %v", trace.DebugReport(err))
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func createAlert(client *kapacitor.Client, spec []byte) error {
+	if len(bytes.TrimSpace(spec)) == 0 {
+		return trace.NotFound("empty configurtion")
+	}
+
+	var alert alert
+	err := yaml.Unmarshal(spec, &alert)
+	if err != nil {
+		return trace.Wrap(err, "failed to unmarshal %s", spec)
+	}
+
+	err = client.CreateAlert(alert.Name, alert.Spec.Formula)
+	if err != nil {
+		return trace.Wrap(err, "failed to create task")
```

Comment:
Same, would be nice to at least get a hint which configmap/secret triggered this error, if possible.

Reply:
Added more context to the caller.